File: python/student_cwa.py
import os
import json

class Exams:
    results = {}
    courses_marks= {}
    courses = []
    
    def __init__(self, name):
        self.name = name
        self.loadResults()

#function to ask user for names of courses 
    def setCourses(self, num_of_courses):
        for course in range(num_of_courses):
            self.courses.append(input(f"Enter the name of course {course+1}: "))
#function to ask user for marks of courses
    def setMarks(self, level, num_of_courses):
        isSet = False
        if level in self.results:
            isSet = True
            print("sorry, the marks for the semester have already been set.")
        if not isSet:
            print("let's first set the courses for the semester")
            self.setCourses(num_of_courses)
            course = 0
            for _ in self.courses:
                mark_hours = []
                mark_hours.append(float(input(f"Enter marks for {self.courses[course]}: ")))
                mark_hours.append(int(input(f"Enter the credit hours for {self.courses[course]}: ")))
                self.courses_marks[self.courses[course]] = mark_hours
                course += 1
            self.results[level] = self.courses_marks
#function to display scores for a particular level
    def getScores(self, level):
        print(" Course \t\t Mark")
        for lev, courses in self.results.items():
            if int(lev) == level:
                for course, mark in courses.items():
                    print(f" {course} \t\t {mark[0]}")

#function that returns the weighted average of the student in a particular level
    def getWA(self, level):
        sum = 0
        total_credit_hours = 0
        for lev, courses in self.results.items():
            if int(lev) == level:
                for course, mark in courses.items():
                    credit_hours = mark[1]
                    sum += mark[0] * credit_hours
                    total_credit_hours += credit_hours
        average = sum/total_credit_hours
        return average

#function that returns the cummulated weighted average of the student's submitted semesters
    def getCWA(self):
        sum = 0
        total_average = 0
        for level in self.results:
            sum += self.getWA(int(level))
            total_average += 1
        return sum/total_average
            
#function to print all the results of the students to the screen
    def showResult(self):
        for level, scores in self.results.items():
            print("Level = level + semester number")
            print(f"Level: {level}:\n Course \t\t Mark")
            for course, mark in scores.items():
                print(f" {course} \t\t {mark[0]}")

#function to save results in a .txt file according to the student's name
    def saveResults(self):
        print("saving file...")
        with open(f"{self.name}.json", "w") as file:
            json.dump(self.results, file, indent = 6)
        print(f"{self.name}.json saved succussfully...")

#funtion to load results in a .txt file according to the student's name
    def loadResults(self):
        print("loading file...")
        if os.path.exists(f"{self.name}.json"):
            with open(f"{self.name}.json") as file:
                self.results = json.load(file)
            print(f"{self.name}.json loaded successfully...")
        else:
            print(f"Sorry, there's no such file with the name {self.name}.json in the directory")

#main function that will be called automatically when the program is run
def main():
    name = input("Student's name: ")
    #create an object instance of the class Student
    student = Exams(name)
    print(f"{name}'s File")
    #commands
    exit, set_marks, get_scores, get_wa, get_cwa, show_results, save_results, load_results=0,1,2,3,4,5,6,7
    #set a forever running loop until break is encoutered
    while True: 
        print("input command:\n 1. Set marks\n 2. Get scores\n 3. Get wa\n 4. Get CWA \n 5. Show results\n 6. Save results\n 7. Load results\n 0. Exit")
        cmd = int(input("Enter command number: "))
        
        if cmd == set_marks or cmd == get_scores or cmd == get_wa:
            year = int(input("Level: "))
            sem = int(input("Semester: "))
            level = sem + year
            if cmd == set_marks:
                num_of_courses = int(input("Number of courses in that semester: "))
                student.setMarks(level, num_of_courses)
            if cmd == get_scores:
                student.getScores(level)
            if cmd == get_wa:
                print(student.getWA(level))
        
        if cmd == get_cwa:
            print(student.getCWA())

        if cmd == show_results:
            student.showResult()

        if cmd == save_results:
            student.saveResults()

        if cmd == load_results:
            student.loadResults()
        if cmd == exit:
            break

if __name__ == "__main__":
    main()

File: cpp/practice/practice/student_cwa.cpp
#include <iostream>
#include <vector>
#include <map>
#include <fstream>

using namespace std;

//define the Exams class
class Exams
{
public:
	//functions to be used outside the class
	Exams(string name);
	void setMarks(int level, int numOfCourses);
	double getWA(int level);
	double getCWA();
	void showResults();
	void saveResults();
	void loadResults();

private:
	string name;
	//funtions and variables only available to the class
	void setCourses(int numOfCourses);
	vector<string> courses;
	map<string, vector<double>> courses_marks;
	map <int, map<string, vector<double>>> results;
};

Exams::Exams(string name) {	
}
void Exams::setCourses(int numOfCourses) {
	for (int course = 1; course <= numOfCourses; course++) {
		string name;
		cout << "Enter the name of course " << course << ": ";
		cin >> name;
		courses.push_back(name);
	}
}

//create setMarks function to recieve the marks for the semester
void Exams::setMarks(int level, int numOfCourses) {
	bool isSet = false;
	//loop through submitted results and check if the particular level has already been set
	//if true, change isSet = true
	for (auto levels : results) {
		if (level == levels.first) {
			isSet = true;
			cout << "Sorry, the marks have already been set." << endl;
		}
	}
	//if not, call the setCourses() function to set the name of the courses for the
	//semester and enter their marks too
	if (!isSet) {
		cout << "wait, let's set the name of the courses first!" << endl;
		//call the setCourses() function
		setCourses(numOfCourses);
		system("pause");
		cout << "Now let's set the marks and the credit hours for each course." << endl;
		for (int course = 0; course < numOfCourses; course++) {
			vector<double> mark_hours;
			double mark;
			cout << "Enter the marks for " << courses[course] << " : ";
			cin >> mark;
			mark_hours.push_back(mark);
			double hours;
			cout << "Enter the credit hours for " << courses[course] << " : ";
			cin >> hours;
			courses_marks[courses[course]] = mark_hours;
		}
	}
	results[level] = courses_marks;
}
//function to get the weighted average for a level
double Exams::getWA(int level) {
	double sum = 0;
	double total_credit_hours = 0;
	
	for (auto levels : results) {
		if (level == levels.first) {
			for (auto course : levels.second) {
				sum += course.second[0] * course.second[1];
				total_credit_hours += course.second[1];
			}
		}
	}
	return sum / total_credit_hours;
}
//function to get the cummulated weighted average for the submitted semesters
double Exams::getCWA() {
	double sum = 0;
	double total_average = 0;
	for (auto level : results) {
		sum += getWA(level.first);
		total_average++;
	}
	return sum / total_average;
}
//function to display the all the results to the screen
void Exams::showResults() {
	cout << "Level = level + semester number" << endl;
	for (auto level : results){
		cout << "Level: " << level.first << endl;
		cout << "\tCourse\t\tMarks" << endl;
		for (auto course : level.second) {
			cout << "\t" << course.first << "\t\t" << course.second[0]<<endl;
		}
	}
}
//function to save results to a .json file
void Exams::saveResults() {
	cout << "saving file..." << endl;
	ofstream file;
	string filename = name + ".txt";
	file.open(filename);
	for (auto level : results) {
		file << level.first << " : " << endl;
		for (auto course : level.second) {
			file << course.first << " : " << course.second[0] << course.second[1] << endl;
		}
	}
	cout << "file saved";

}
void Exams::loadResults() {
	cout << "loading file..." << endl;
	ifstream file;
	string filename = name + ".txt";
	file.open(filename);

}
//function to load results from a .json file
int main() {
	Exams prince("Prince");
	prince.setMarks(100, 2);
}